### ISSUE TITLE
picom crashes when running some applications

### ISSUE THREAD

**Issue by FrozenArcher** [yshui/compton#1111]
### Platform

Gentoo amd64

### GPU, drivers, and screen setup

Intel(R) HD Graphics 620 (KBL GT2), mesa 23.1.3, one built-in monitor

```
$ glxinfo -B
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Intel (0x8086)
    Device: Mesa Intel(R) HD Graphics 620 (KBL GT2) (0x5916)
    Version: 23.1.3
    Accelerated: yes
    Video memory: 7697MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.6
    Max compat profile version: 4.6
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
OpenGL vendor string: Intel
OpenGL renderer string: Mesa Intel(R) HD Graphics 620 (KBL GT2)
OpenGL core profile version string: 4.6 (Core Profile) Mesa 23.1.3
OpenGL core profile shading language version string: 4.60
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 4.6 (Compatibility Profile) Mesa 23.1.3
OpenGL shading language version string: 4.60
OpenGL context flags: (none)
OpenGL profile mask: compatibility profile

OpenGL ES profile version string: OpenGL ES 3.2 Mesa 23.1.3
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
```

### Environment

dwm

### picom version

```
$ picom --version
v10
```
<details>
<summary>Diagnostics</summary>

**Version:** v10

### Extensions:

* Shape: Yes
* XRandR: Yes
* Present: Present

### Misc:

* Use Overlay: Yes
* Config file used: /home/fracher/.config/picom/picom.conf

### Drivers (inaccurate):

modesetting

### Backend: glx

* Driver vendors:
 * GLX: Mesa Project and SGI
 * GL: Intel
* GL renderer: Mesa Intel(R) HD Graphics 620 (KBL GT2)
* Accelerated: 1

### Backend: egl

* Driver vendors:
 * EGL: Mesa Project
 * EGL driver: iris
 * GL: Intel
* GL renderer: Mesa Intel(R) HD Graphics 620 (KBL GT2)

</details>

### Configuration:
<details>
    <summary>Configuration file</summary>

```
// Paste your configuration here
################################
#           Animations         #
################################

#enable or disable animations
animations = true;
#change animation speed of windows in current tag e.g open window in current tag
animation-stiffness-in-tag = 300;
#change animation speed of windows when tag changes
animation-stiffness-tag-change = 700;

animation-window-mass = 0.7;
animation-dampening = 20;
animation-clamping = false;

#open windows
animation-for-open-window = "zoom";
#minimize or close windows
#animation-for-unmap-window = "squeeze";
#popup windows
animation-for-transient-window = "zoom"; #available options: slide-up, slide-down, slide-left, slide-right, squeeze, squeeze-bottom, zoom

#set animation for windows being transitioned out while changings tags
#animation-for-prev-tag = "minimize";
#enables fading for windows being transitioned out while changings tags
#enable-fading-prev-tag = true;

#set animation for windows being transitioned in while changings tags
#animation-for-next-tag = "slide-in-center";
#enables fading for windows being transitioned in while changings tags
#enable-fading-next-tag = true;

#################################
#             Shadows           #
#################################

# Enabled client-side shadows on windows. Note desktop windows
# (windows with '_NET_WM_WINDOW_TYPE_DESKTOP') never get shadow,
# unless explicitly requested using the wintypes option.
#
# shadow = false
shadow = true;

# The blur radius for shadows, in pixels. (defaults to 12)
# shadow-radius = 12
shadow-radius = 30;

# The opacity of shadows. (0.0 - 1.0, defaults to 0.75)
# shadow-opacity = .75
shadow-opacity = 0.7

# The left offset for shadows, in pixels. (defaults to -15)
# shadow-offset-x = -15
shadow-offset-x = -30;

# The top offset for shadows, in pixels. (defaults to -15)
# shadow-offset-y = -15
shadow-offset-y = -30;

# Red color value of shadow (0.0 - 1.0, defaults to 0).
# shadow-red = 0

# Green color value of shadow (0.0 - 1.0, defaults to 0).
# shadow-green = 0

# Blue color value of shadow (0.0 - 1.0, defaults to 0).
# shadow-blue = 0

# Hex string color value of shadow (#000000 - #FFFFFF, defaults to #000000). This option will override options set shadow-(red/green/blue)
shadow-color = "#000000"

# Specify a list of conditions of windows that should have no shadow.
#
# examples:
#   shadow-exclude = "n:e:Notification";
#
# shadow-exclude = []
shadow-exclude = [
  "class_g = 'pop_report'",
  "class_g = 'Cairo-clock'",
  "class_g = 'awesome'",
  "class_g = 'Xfce4-panel'",
  "!focused",
  "window_type != 'normal'",
  "_GTK_FRAME_EXTENTS@:c"
];

# Specify a list of conditions of windows that should have no shadow painted over, such as a dock window.
# clip-shadow-above = []

# Specify a X geometry that describes the region in which shadow should not
# be painted in, such as a dock window region. Use
#    shadow-exclude-reg = "x10+0+0"
# for example, if the 10 pixels on the bottom of the screen should not have shadows painted on.
#
# shadow-exclude-reg = ""

# Crop shadow of a window fully on a particular Xinerama screen to the screen.
# xinerama-shadow-crop = false


#################################
#           Fading              #
#################################


# Fade windows in/out when opening/closing and when opacity changes,
#  unless no-fading-openclose is used.
fading = true;

# Opacity change between steps while fading in. (0.01 - 1.0, defaults to 0.028)
# fade-in-step = 0.028
fade-in-step = 0.07;

# Opacity change between steps while fading out. (0.01 - 1.0, defaults to 0.03)
# fade-out-step = 0.03
fade-out-step = 0.07;

# The time between steps in fade step, in milliseconds. (> 0, defaults to 10)
fade-delta = 16

# Specify a list of conditions of windows that should not be faded.
# fade-exclude = []

# Do not fade on window open/close.
# no-fading-openclose = false

# Do not fade destroyed ARGB windows with WM frame. Workaround of bugs in Openbox, Fluxbox, etc.
# no-fading-destroyed-argb = false


#################################
#   Transparency / Opacity      #
#################################


# Opacity of inactive windows. (0.1 - 1.0, defaults to 1.0)
inactive-opacity = 0.9

# Opacity of window titlebars and borders. (0.1 - 1.0, disabled by default)
# frame-opacity = 1.0

# Let inactive opacity set by -i override the '_NET_WM_WINDOW_OPACITY' values of windows.
# inactive-opacity-override = true
inactive-opacity-override = true;

# Default opacity for active windows. (0.0 - 1.0, defaults to 1.0)
# active-opacity = 1.0

# Dim inactive windows. (0.0 - 1.0, defaults to 0.0)
inactive-dim = 0.2

# Specify a list of conditions of windows that should never be considered focused.
# focus-exclude = []
focus-exclude = [
  "class_g = 'Cairo-clock'" ,
  "class_g = 'Dunst'",
  "window_type = 'notification'",
  "window_type = 'dock'",
  "class_g = 'pop_report'",
  "class_g = 'awesome'",
  "class_g = 'Xfce4-panel'",
];

# Use fixed inactive dim value, instead of adjusting according to window opacity.
# inactive-dim-fixed = 1.0

#################################
#           Corners             #
#################################

# Sets the radius of rounded window corners. When > 0, the compositor will
# round the corners of windows. Does not interact well with
# `transparent-clipping`.
corner-radius = 10;

# Exclude conditions for rounded corners.
rounded-corners-exclude = [
  "window_type = 'tooltip'",
  "window_type = 'popup_menu'",
  "window_type = 'dropdown_menu'",
  "window_type = 'toolbar'",
  "window_type = 'utility'",
  "window_type = 'splash'",
  "window_type = 'dialog'",
  "window_type = 'dnd'",
  "window_type = 'desktop'",
  "class_g = 'eww-music'",
  "class_g = 'eww-calendar'",
  "class_g = 'Rofi'",
  "class_g = 'i3lock'",
  "class_g = 'awesome'",
  "class_g = 'Xfce4-panel'",
  "_NET_WM_STATE@:32a *= '_NET_WM_STATE_FULLSCREEN'",
];

blur: {
  method = "dual_kawase";
  strength = 4;
  background = true;
  background-frame = true;
  background-fixed = true;
}


# Exclude conditions for background blur.
# blur-background-exclude = []
blur-background-exclude = [
  "window_type != 'normal'",
  "class_g = 'Chromium'",
  "class_g = 'Discord'",
  "class_g = 'Peek'",
  "class_g = 'flameshot'",
  "class_g = 'xdg-desktop-portal-gnome'",
  "_GTK_FRAME_EXTENTS@:c",
];

#################################
#       General Settings        #
#################################

# Daemonize process. Fork to background after initialization. Causes issues with certain (badly-written) drivers.
# daemon = false

# Specify the backend to use: `xrender`, `glx`, `egl` or `xr_glx_hybrid`.
# `xrender` is the default one.
#
backend = "glx"

# Use higher precision during rendering, and apply dither when presenting the
# rendered screen. Reduces banding artifacts, but might cause performance
# degradation. Only works with OpenGL.
dithered-present = false;

# Enable/disable VSync.
vsync = true

# Enable remote control via D-Bus. See the *D-BUS API* section below for more details.
# dbus = false

# Try to detect WM windows (a non-override-redirect window with no
# child that has 'WM_STATE') and mark them as active.
#
# mark-wmwin-focused = false
mark-wmwin-focused = true;

# Mark override-redirect windows that doesn't have a child window with 'WM_STATE' focused.
# mark-ovredir-focused = false
mark-ovredir-focused = true;

# Try to detect windows with rounded corners and don't consider them
# shaped windows. The accuracy is not very high, unfortunately.
#
# detect-rounded-corners = false
detect-rounded-corners = false;

# Detect '_NET_WM_WINDOW_OPACITY' on client windows, useful for window managers
# not passing '_NET_WM_WINDOW_OPACITY' of client windows to frame windows.
#
# detect-client-opacity = false
detect-client-opacity = false;

# Use EWMH '_NET_ACTIVE_WINDOW' to determine currently focused window,
# rather than listening to 'FocusIn'/'FocusOut' event. Might have more accuracy,
# provided that the WM supports it.
#
use-ewmh-active-win = false;

# Unredirect all windows if a full-screen opaque window is detected,
# to maximize performance for full-screen windows. Known to cause flickering
# when redirecting/unredirecting windows.
#
unredir-if-possible = true;
unredir-if-possible-exclude = [
  "class_g = 'Brave-browser'",
  "class_g = 'vlc'",
  "class_g = 'i3lock'",
  "class_g = 'mpv'",
  "class_g = 'Discord'",
  "name *= 'eww'",
]


# Delay before unredirecting the window, in milliseconds. Defaults to 0.
# unredir-if-possible-delay = 0

# Conditions of windows that shouldn't be considered full-screen for unredirecting screen.
# unredir-if-possible-exclude = []

# Use 'WM_TRANSIENT_FOR' to group windows, and consider windows
# in the same group focused at the same time.
#
# detect-transient = false
detect-transient = true;

# Use 'WM_CLIENT_LEADER' to group windows, and consider windows in the same
# group focused at the same time. This usually means windows from the same application
# will be considered focused or unfocused at the same time.
# 'WM_TRANSIENT_FOR' has higher priority if detect-transient is enabled, too.
#
# detect-client-leader = false

# Resize damaged region by a specific number of pixels.
# A positive value enlarges it while a negative one shrinks it.
# If the value is positive, those additional pixels will not be actually painted
# to screen, only used in blur calculation, and such. (Due to technical limitations,
# with use-damage, those pixels will still be incorrectly painted to screen.)
# Primarily used to fix the line corruption issues of blur,
# in which case you should use the blur radius value here
# (e.g. with a 3x3 kernel, you should use `--resize-damage 1`,
# with a 5x5 one you use `--resize-damage 2`, and so on).
# May or may not work with *--glx-no-stencil*. Shrinking doesn't function correctly.
#
# resize-damage = 1

# Specify a list of conditions of windows that should be painted with inverted color.
# Resource-hogging, and is not well tested.
#
# invert-color-include = []

# GLX backend: Avoid using stencil buffer, useful if you don't have a stencil buffer.
# Might cause incorrect opacity when rendering transparent content (but never
# practically happened) and may not work with blur-background.
# My tests show a 15% performance boost. Recommended.
#
glx-no-stencil = true;

# GLX backend: Avoid rebinding pixmap on window damage.
# Probably could improve performance on rapid window content changes,
# but is known to break things on some drivers (LLVMpipe, xf86-video-intel, etc.).
# Recommended if it works.
#
# glx-no-rebind-pixmap = false

# Disable the use of damage information.
# This cause the whole screen to be redrawn every time, instead of the part of the screen
# has actually changed. Potentially degrades the performance, but might fix some artifacts.
# The opposing option is use-damage
#
# no-use-damage = false
use-damage = true;

# Use X Sync fence to sync clients' draw calls, to make sure all draw
# calls are finished before picom starts drawing. Needed on nvidia-drivers
# with GLX backend for some users.
#
xrender-sync-fence = true;

# GLX backend: Use specified GLSL fragment shader for rendering window contents.
# See `compton-default-fshader-win.glsl` and `compton-fake-transparency-fshader-win.glsl`
# in the source tree for examples.
#
window-shader-fg = "default";

# Force all windows to be painted with blending. Useful if you
# have a glx-fshader-win that could turn opaque pixels transparent.
#
# force-win-blend = true;

# Do not use EWMH to detect fullscreen windows.
# Reverts to checking if a window is fullscreen based only on its size and coordinates.
#
no-ewmh-fullscreen = true

# Dimming bright windows so their brightness doesn't exceed this set value.
# Brightness of a window is estimated by averaging all pixels in the window,
# so this could comes with a performance hit.
# Setting this to 1.0 disables this behaviour. Requires --use-damage to be disabled. (default: 1.0)
#
# max-brightness = 1.0

# Make transparent windows clip other windows like non-transparent windows do,
# instead of blending on top of them.
#
transparent-clipping = false;

# Specify a list of conditions of windows that should never have transparent
# clipping applied. Useful for screenshot tools, where you need to be able to see through transparent parts of the window.
#
# transparent-clipping-exclude = []

# Set the log level. Possible values are:
#  "trace", "debug", "info", "warn", "error"
# in increasing level of importance. Case doesn't matter.
# If using the "TRACE" log level, it's better to log into a file
# using *--log-file*, since it can generate a huge stream of logs.
#
# log-level = "debug"
log-level = "warn";

# Set the log file.
# If *--log-file* is never specified, logs will be written to stderr.
# Otherwise, logs will to written to the given file, though some of the early
# logs might still be written to the stderr.
# When setting this option from the config file, it is recommended to use an absolute path.
#
# log-file = "/path/to/your/log/file"

# Show all X errors (for debugging)
# show-all-xerrors = false

# Write process ID to a file.
# write-pid-path = "/path/to/your/log/file"

# Window type settings
#
# 'WINDOW_TYPE' is one of the 15 window types defined in EWMH standard:
#     "unknown", "desktop", "dock", "toolbar", "menu", "utility",
#     "splash", "dialog", "normal", "dropdown_menu", "popup_menu",
#     "tooltip", "notification", "combo", and "dnd".
#
# Following per window-type options are available: ::
#
#   fade, shadow:::
#     Controls window-type-specific shadow and fade settings.
#
#   opacity:::
#     Controls default opacity of the window type.
#
#   focus:::
#     Controls whether the window of this type is to be always considered focused.
#     (By default, all window types except "normal" and "dialog" has this on.)
#
#   full-shadow:::
#     Controls whether shadow is drawn under the parts of the window that you
#     normally won't be able to see. Useful when the window has parts of it
#     transparent, and you want shadows in those areas.
#
#   clip-shadow-above:::
#     Controls whether shadows that would have been drawn above the window should
#     be clipped. Useful for dock windows that should have no shadow painted on top.
#
#   redir-ignore:::
#     Controls whether this type of windows should cause screen to become
#     redirected again after been unredirected. If you have unredir-if-possible
#     set, and doesn't want certain window to cause unnecessary screen redirection,
#     you can set this to `true`.
#
wintypes:
{
  normal = { full-shadow = false; }
  tooltip = { fade = true; shadow = true; opacity = 1.0; focus = true; full-shadow = true; };
  dock = { full-shadow = true; clip-shadow-above = true; animation = "slide-down" }
  dnd = { shadow = false; }
  popup_menu = { opacity = 1.0; shadow = false; }
  dropdown_menu = { opacity = 1.0; shadow = false; }
  notification = { full-shadow = true }
};
```
</details>

### Steps of reproduction

1. $ killall picom
2. $ picom -b
3. start some applications (especially those based on electron, e.g. vscode)
4. picom crashes.

### Expected behavior

picom continues to work

### Current Behavior

picom crashes with message:

```
Xlib: ignoring invalid extension event 161
Xlib: ignoring invalid extension event 161
Xlib: ignoring invalid extension event 161
Xlib: ignoring invalid extension event 161
Xlib: ignoring invalid extension event 161
Xlib: ignoring invalid extension event 161
Xlib: ignoring invalid extension event 161
Xlib: ignoring invalid extension event 161
Xlib: ignoring invalid extension event 161
Xlib: ignoring invalid extension event 161
picom: ../picom-10.2/src/win.c:349: win_bind_mask: Assertion `!w->mask_image' failed.
[1]    71949 IOT instruction (core dumped)  picom
```

**Comment by absolutelynothelix** [comment]
try picom built from the latest commit in the next branch.

**Comment by absolutelynothelix** [comment]
closing as stale.

**Comment by haarp** [comment]
fwiw, I'm also seeing this issue. Gentoo Linux, Xfce4, picom-10.2, (#1091 applied to fix another crash)

```
glxinfo -B
name of display: :0.0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: AMD (0x1002)
    Device: AMD Radeon Graphics (rembrandt, LLVM 16.0.6, DRM 3.49, 6.1.53-gentoo-r1-HH) (0x1681)
    Version: 23.1.8
    Accelerated: yes
    Video memory: 2048MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 4.6
    Max compat profile version: 4.6
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
Memory info (GL_ATI_meminfo):
    VBO free memory - total: 520 MB, largest block: 520 MB
    VBO free aux. memory - total: 13631 MB, largest block: 13631 MB
    Texture free memory - total: 520 MB, largest block: 520 MB
    Texture free aux. memory - total: 13631 MB, largest block: 13631 MB
    Renderbuffer free memory - total: 520 MB, largest block: 520 MB
    Renderbuffer free aux. memory - total: 13631 MB, largest block: 13631 MB
Memory info (GL_NVX_gpu_memory_info):
    Dedicated video memory: 2048 MB
    Total available memory: 16975 MB
    Currently available dedicated video memory: 520 MB
OpenGL vendor string: AMD
OpenGL renderer string: AMD Radeon Graphics (rembrandt, LLVM 16.0.6, DRM 3.49, 6.1.53-gentoo-r1-HH)
OpenGL core profile version string: 4.6 (Core Profile) Mesa 23.1.8
OpenGL core profile shading language version string: 4.60
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 4.6 (Compatibility Profile) Mesa 23.1.8
OpenGL shading language version string: 4.60
OpenGL context flags: (none)
OpenGL profile mask: compatibility profile

OpenGL ES profile version string: OpenGL ES 3.2 Mesa 23.1.8
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
```

It seems to happen randomly, possibly when starting apps or after unlocking the screenlocker. Can't say for sure when it crashes, but it happens a couple of times per week.

Right now I'm testing newest git (`vgit-751f3`) and will report back in a week or so.

**Comment by haarp** [comment]
Haven't had any more crashes, looks like it's fixed as of at least 751f3. :D

**Comment by tofurky** [comment]
i was also seeing frequent crashes with the following on glx backend on ubuntu 23.10 (10.2):
```
picom: ../src/win.c:349: win_bind_mask: Assertion `!w->mask_image' failed.
```

so far so good after switching to tip of next last night. i'll be sure to report back if i see further crashes.

**Comment by Monsterovich** [comment]
> i was also seeing frequent crashes with the following on glx backend on ubuntu 23.10 (10.2):
> 
> ```
> picom: ../src/win.c:349: win_bind_mask: Assertion `!w->mask_image' failed.
> ```
> 
> so far so good after switching to tip of next last night. i'll be sure to report back if i see further crashes.

Please, try [this](https://github.com/yshui/picom/pull/1091) patch.

**Comment by haarp** [comment]
> Please, try [this](https://github.com/yshui/picom/pull/1091) patch.

In my experience, this fixes a different crash bug. I had both these bugs.

`Stale flags set for mapped window 0x0123456789 during backend destruction` is fixed by your patch, `win_bind_mask: Assertion '!w->mask_image' failed.` is fixed by something between v10.2 and current `next` branch.
